### ISSUE TITLE
No error: False-positive Control Flow with getters

### ISSUE THREAD

**Issue by EugeneZee** [Microsoft/TypeScript#12140]
**TypeScript Version:**  2.0.8

**Code**

```ts
class A {
	public a: number | A;

	public get b(): number | A {
		this.a = new A();
		return this.a;
	}

	public c(): number | A {
		this.a = new A();
		return this.a;
	}
}

const obj = new A();

if (typeof obj.a === 'number') {
	const n = obj.a++;                           // NO ERROR - CORRECT
}

if (typeof obj.b === 'number') {
	const n: number = obj.b;                   // NO ERROR - WRONG
}

if (typeof obj.c() === 'number') {
	const n: number = obj.c();                // ERROR - CORRECT
}
```

**Expected behavior:**
Error with `obj.b` because type of getters results can change after each call.

**Actual behavior:** 
`obj.b` is false treated as a `number`.

**Comment by HerringtonDarkholme** [comment]
Hi, though not written in specs, getters are probably treated as plain property access in TypeScript(and therefore side-effect free). There is no way to tell a getter from a property on type level.

So this issue is by design.

**Comment by RyanCavanaugh** [comment]
This is correct - TypeScript assumes properties don't mutate on their own.

**Comment by EugeneZee** [comment]
Obviously this assumption is wrong, but you know better what can be and what cannot be implemented in TS.
Taking into account that this is a limitation I hope that sometime we'll have const methods in TS where it will be prohibited to modify state of an object.
